### ISSUE TITLE
Cradle reloads three times before correctly setting a typechecked module.

### ISSUE THREAD

**Issue by fendor** [mpickering/haskell-ide-engine#10]
Weird behaviour when loading a cradle.

On a test project, I can observe the following behaviour:

<details>
<summary>Logs</summary>

```
2019-09-25 15:29:50.543331392 [ThreadId 4] - run entered for hie-wrapper(hie-wrapper) Version 1.0.0.0 x86_64 ghc-8.6.4
2019-09-25 15:29:50.571136909 [ThreadId 4] - Current directory:/home/baldr/Documents/definition
2019-09-25 15:29:50.615901821 [ThreadId 4] - Operating system:linux
2019-09-25 15:29:50.62436447 [ThreadId 4] - Cradle directory:/home/baldr/Documents/definition
2019-09-25 15:29:50.625392815 [ThreadId 4] - Using stack GHC version
2019-09-25 15:29:53.714345915 [ThreadId 4] - Project GHC version:8.6.4
2019-09-25 15:29:53.714614318 [ThreadId 4] - hie exe candidates :["hie-8.6.4","hie-8.6","hie"]
2019-09-25 15:29:53.715424774 [ThreadId 4] - found hie exe at:/home/baldr/.local/bin/hie-8.6.4
2019-09-25 15:29:53.715824582 [ThreadId 4] - args:[]
2019-09-25 15:29:53.716023549 [ThreadId 4] - launching ....



2019-09-25 15:29:53.718588811 [ThreadId 4] - Using stack GHC version
2019-09-25 15:29:55.816193778 [ThreadId 4] - Run entered for HIE(hie-8.6.4) Version 1.0.0.0 x86_64 ghc-8.6.4
2019-09-25 15:29:55.816359912 [ThreadId 4] - Current directory:/home/baldr/Documents/definition
2019-09-25 15:29:55.825018211 [ThreadId 9] - Using stack GHC version
Using hie version: Version 1.0.0.0 x86_64 ghc-8.6.4
("lookupCradle","/home/baldr/Documents/definition/Bar.hs",GhcModuleCache {cradleCache = Data.Trie.fromList [], uriCaches = fromList [], currentCradle = Nothing})
("New cradle","/home/baldr/Documents/definition/Bar.hs")
Cradle {cradleRootDir = "/home/baldr/Documents/definition", cradleOptsProg = CradleAction: stack}
Using hoogle db at: /home/baldr/.hoogle/default-haskell-5.0.17.hoo
[]
("lookupCradle","/home/baldr/Documents/definition/Bar.hs",GhcModuleCache {cradleCache = Data.Trie.fromList [], uriCaches = fromList [], currentCradle = Just ([],Cradle {cradleRootDir = "/home/baldr/Documents/definition", cradleOptsProg = CradleAction: stack})})
("just","/home/baldr/Documents/definition/Bar.hs",[])
("New cradle","/home/baldr/Documents/definition/Bar.hs")
Cradle {cradleRootDir = "/home/baldr/Documents/definition", cradleOptsProg = CradleAction: stack}
[]
("lookupCradle","/home/baldr/Documents/definition/Bar.hs",GhcModuleCache {cradleCache = Data.Trie.fromList [], uriCaches = fromList [], currentCradle = Just ([],Cradle {cradleRootDir = "/home/baldr/Documents/definition", cradleOptsProg = CradleAction: stack})})
("just","/home/baldr/Documents/definition/Bar.hs",[])
("New cradle","/home/baldr/Documents/definition/Bar.hs")
Cradle {cradleRootDir = "/home/baldr/Documents/definition", cradleOptsProg = CradleAction: stack}
[]
loadFile:2 /home/baldr/Documents/definition
loadFile:3 0
setTargets
  (/home/baldr/Documents/definition/Bar.hs , /tmp/VFS28381-0.hs)
  [*/tmp/VFS28381-0.hs]
modGraph
  [ModSummary {
      ms_hs_date = 2019-09-25 13:30:14.198465633 UTC
      ms_mod = Bar,
      ms_textual_imps = [(Nothing, Prelude)]
      ms_srcimps = []
   }]
modGraph
  [ModLocation {ml_hs_file = Just "/tmp/VFS28381-0.hs", ml_hi_file = "/home/baldr/.cache/haskell-ide-engine/32e16f78874274cbb4f10898157ab2d191388b6b/Bar.hi", ml_obj_file = "/home/baldr/Documents/definition/.stack-work/odir/Bar.o"}]
hidir
  Just [/, h, o, m, e, /, b, a, l, d, r, /, ., c, a, c, h, e, /, h,
        a, s, k, e, l, l, -, i, d, e, -, e, n, g, i, n, e, /, 3, 2, e, 1,
        6, f, 7, 8, 8, 7, 4, 2, 7, 4, c, b, b, 4, f, 1, 0, 8, 9, 8, 1, 5,
        7, a, b, 2, d, 1, 9, 1, 3, 8, 8, b, 6, b]
loading (1, 1)
"Use hook"
loaded
  /home/baldr/Documents/definition/Bar.hs
  /tmp/VFS28381-0.hs
("tms",[Just "/tmp/VFS28381-0.hs"])
Diags
  [(NormalizedUri "file:///home/baldr/Documents/definition/Bar.hs",[])]
```

</details>

What weird is, that it tries to load the cradle 3 times before it actually can be loaded and produce a module graph.
This should probably not happen, any idea what is going on?

EDIT: I think what is going on is that after loading the cradle, the module is getting typechecked. 
When something requests diagnostics or something like that and needs a typechecked module, the module will be typechecked within `runWithContext`, which tries to load again the cradle.

**Comment by fendor** [comment]
This issue might be stack related.

**Comment by fendor** [comment]
The problem seems to be that "stack repl" does not return the targets of a project. 
Work around is currently to add the FilePath, which we load the cradle for, to the compiler options if the cradle type is "stack".

**Comment by fendor** [comment]
Currently solved by #8
